### ISSUE TITLE
SseEmitter (ResponseBodyEmitter) keeps collecting messages if not initialized

### ISSUE THREAD

**Issue by kgcsabi** [spring-projects/spring-framework#25442]
### Details

**Component:** spring-webmvc
**Affects:** 5.2.7.RELEASE
**Impact:** Memory leak

### Description

The ResponseBodyEmitter instance is normally initialized by the framework upon returned by the RestController method.
The messages that are passed in the meantime are collected in a LinkedHashSet.

When the connection is dropped by the client early enough then the **initialization will never be invoked**.

These instances keep collecting the messages with **no limitation**.
To make matters worse, these instances are **not evicted when timeout** is reached.

This may result in OutOfMemoryError on the long run.

### Steps to reproduce:

1. Start Server app in debug mode
2. Add breakpoint to org.springframework.web.servlet.mvc.method.annotation.ResponseBodyEmitterReturnValueHandler:129 //emitter.initialize(handler);
3. Start Client app
4. When breakpoint hit then stop Client app

### Attachments

- [sse.zip](https://github.com/spring-projects/spring-framework/files/4952321/sse.zip)
- [sse-with-timeout.log](https://github.com/spring-projects/spring-framework/files/4952325/sse-with-timeout.log)

### Ideas

In my opinion this may be addressed by
- limiting the time waiting for initialization
- limiting the number of messages kept before initialization
- a combination of the above
- a custom handler for early send attempts

**Comment by IstvanN** [comment]
+1

**Comment by rstoyanchev** [comment]
I can't reproduce the exact scenario. The ISE "Async request already returned to container" from the logs is under `WebAsyncRequest#startDeferredResultProcessing` which is a couple of lines before `emitter.initialize(handler)`. That means the connection must have been closed by then. I tried  dropping the connection at that stage but I still can't get `startDeferredResultProcessing` to raise the same error. 

In any case it's clear that we need to protect against very early connection issues.
